libraries/standard/mqtt/src/mqtt_lightweight.c

```diff
@@ -192,6 +247,508 @@ static int32_t recvExact( MQTTTransportRecvFunc_t recvFunc,
+        if( remainingLength < qos0Minimum )
+        {
+            IotLogDebugWithArgs( "QoS 0 PUBLISH cannot have a remaining length less than %lu.",
+                                 qos0Minimum );
+
+            status = MQTTBadResponse;
+        }
+    }
+    else
+    {
+        /* Check that the "Remaining length" is greater than the minimum. For
+         * QoS 1 or 2, this will be two bytes greater than for QoS 0 due to the
+         * packet identifier. */
+        if( remainingLength < ( qos0Minimum + 2U ) )
+        {
+            IotLogDebugWithArgs( "QoS 1 or 2 PUBLISH cannot have a remaining length less than %lu.",
+                                 qos0Minimum + 2U );
+
+            status = MQTTBadResponse;
+        }
+    }
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t _processPublishFlags( uint8_t publishFlags, MQTTPublishInfo_t * const pPublishInfo )
+{
+    MQTTStatus_t status = MQTTSuccess;
+
+    assert( pPublishInfo != NULL );
+
+    /* Check for QoS 2. */
+    if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_QOS2 ) )
+    {
+        /* PUBLISH packet is invalid if both QoS 1 and QoS 2 bits are set. */
+        if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_QOS1 ) )
+        {
+            IotLogDebug( "Bad QoS: 3." );
+
+            status = MQTTBadResponse;
+        }
+        else
+        {
+            pPublishInfo->qos = MQTTQoS2;
+        }
+    }
+    /* Check for QoS 1. */
+    else if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_QOS1 ) )
+    {
+        pPublishInfo->qos = MQTTQoS1;
+    }
+    /* If the PUBLISH isn't QoS 1 or 2, then it's QoS 0. */
+    else
+    {
+        pPublishInfo->qos = MQTTQoS0;
+    }
+
+    if( status == MQTTSuccess )
+    {
+        IotLogDebugWithArgs( "QoS is %d.", pPublishInfo->qos );
+
+        /* Parse the Retain bit. */
+        pPublishInfo->retain = UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_RETAIN );
+
+        IotLogDebugWithArgs( "Retain bit is %d.", pPublishInfo->retain );
+
+        /* Parse the DUP bit. */
+        if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_DUP ) )
+        {
+            IotLogDebug( "DUP is 1." );
+        }
+        else
+        {
+            IotLogDebug( "DUP is 0." );
+        }
+    }
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t _deserializeConnack( const MQTTPacketInfo_t * const pConnack,
+                                         bool * const pSessionPresent )
+{
+    MQTTStatus_t status = MQTTSuccess;
+
+    assert( pConnack != NULL );
+    assert( pSessionPresent != NULL );
+    const uint8_t * pRemainingData = pConnack->pRemainingData;
+
+    /* If logging is enabled, declare the CONNACK response code strings. The
+     * fourth byte of CONNACK indexes into this array for the corresponding response. */
+    #if LIBRARY_LOG_LEVEL > IOT_LOG_NONE
```

Comment:
Should we create a function here instead, which will not be linked if logging is disabled. ?

Reply:
Resolving since it has been moved inside a function.

Reply:
Looks like we need a "WHY" comment here. A comment explaining that this is in a function so that it can be removed from the build easily would help cusotmers ecplain the code.

Reply:
Added a comment explaining the block